Switch VariantEval to MultiVariantWalkerGroupedOnStart

bbimber opened the pull request:
@cmnbroad This is related to discussion on issue 5439. This is not a final product yet. I'm opening the PR to see how it works on travis and to push discussion here.

This PR is not trying to fix all issues with VariantEval. It's trying to address these:

1) Switch to MultiVariantWalkerGroupedOnStart, primarily to avoid the constant re-querying of variants per site that took place in VariantEvalUtils.bindVariantContexts(). I believe this will substantially reduce the number of instance in which featureContext.getValues() is called.

2) I tried to move, but not full fix, some of the tight linkage between the VariantEval Walker class and the plugin classes. I also made a VariantEvalArgumentCollection to start separating these. Toward this objective, this PR does: a) makes a VariantEvalContext class, which is what gets passed to the VariantStratifier classes, and b) I try to reduce exposing the walker class directly to VariantStratifier and VariantEvaluator. The latter is not completely done, but I think this is moving it in that direction. At several points I stopped for the sake of keeping changes in one PR manageable.

bbimber commented:
There is going to be a behavior different between the old iteration pattern. The existing code will iterate variants to establish whitelist of start sites, and then re-query variants overlapping those starts. This is generally identical in practice to MultiVariantWalkerGroupedOnStart; however, when -L is supplied, it can give differences. I'm not sure which I think is correct. 

Example, from VariantEvalIntegrationTest.testFunctionClassWithSnpeff():

        // The DbSNP VCF has two variants:
        // 1       1423281 rs374004343     GGAAGC  G       .       .
        // 1       1423281 rs79849353      G       A       .       .
        // If we use 1423281 as the interval, we find both.

        // The SnpEff VCF has these two:
        // 1       1423282 .       GAAGC   G
        // 1       1423286 .       C       G

The SnpEff VCF is provided as -L when running the tool. Both SnpEff and DbSnp will be used as drivingVariants.  Therefore when we hit interval 1423282, the DbSnp variant 1423281:GGAAGC>G variant will be included, but not 1423281:G>A.

When using MultiVariantWalkerGroupedOnStart, this means only that variant is passed downstream.

Previously, VariantEval would simply establish the whitelist of start sites (meaning the lowest from the group, or 1423281) and iterate them. This means that even though 1423282 is the interval from -L, it picks up the overlapping 1423281:GGAAGC>G, which effectively makes 1423281 a whitelisted start site. The existing behavior would iterate each start site and re-query overlapping variants in VariantEvalUtils.bindVariantContexts(). It would call:

List<VariantContext> prior = featureContext.getValues(track, referenceContext.getInterval().getStart());

In this instance, it would query on start=1423281, meaning it picked up both DbSnp sites, even though 1423281:G>A is not within the intervals supplied by -L.

This is sort of a fringe case. I dont see how to fix this without reintroducing the expensive re-query step. I'm currently thinking that the existing behavior is probably what's wrong, but I'm going to consider this a little more. My last commit highlights this case

bbimber commented:
@cmnbroad I did a pass with a little cleanup. I think this is ready for review, but has a couple things I left the might help review:

1) As noted above, the primary purpose here is to migrate to MultiVariantWalkerGroupedOnStart, and remove the redundant re-querying of comp alleles. This seems to work, but has the effect of altering behavior in some cases, described more above.  In VariantEvalUtils.java I left some debugging code that illustrates the behavior difference that will occur. 

2) It is a fair question as to whether changing the behavior of what is or isnt considered an overlap is appropriate. For now I'm making changes as though it is, since it's sort of a fringe case and this is a beta tool, but it should be discussed.

3) There are ~6 tests with altered expectations, due to that change in detecting overlaps. I just checked in their updated expectations, since it helps illustrate how the iteration change would impact results

cmnbroad commented:
Hm. At first I thought that the old behavior was wrong, but after more consideration, I think it was correct. It does seem that the `1423282 .       GAAGC   G` and `1423281 rs374004343     GGAAGC  G` variants should be processed together, since they both overlap `1423282`.

bbimber commented:
@cmnbroad yes, 1423282 . GAAGC G and 1423281 rs374004343 GGAAGC G should be processed together, and they still are here. that behavior is not changed.

the differences related to the dbSNP variant 1423281 rs79849353      G > A. That variant does not overlap the eval VCF and it does not fall within the -L intervals.  I would argue that it should not be considered overlapping, since it doesnt span 1423282. The code in this PR will omit it.

cmnbroad commented:
Ah, I see - thats sounds less ominous. Some of the expected results appear to have changed quite a lot though.

Can you rebase this on current master (much preferred over merging master into this branch since it keeps your commits together) - there are conflicts with a recent set of changes we merged in last week. Note that some of those changes are in code that you moved (i.e., into the new ArgumentCollection class) so they'll have to be propagated to the new code. 

Also, the (non VariantEval) test failures in this branch are fallout from the change to add the source to the VariantContexts. I think @jonn-smith/@kclydon have a fix for that so the stalled https://github.com/broadinstitute/gatk/pull/4571 can be merged, since that will be a prerequisite for this branch.

bbimber commented:
@cmnbroad yes, i saw the conflict and will get to that and rebase.

regarding test expectation changes: in a couple instances the genome territory number changes quite a lot. this is because I needed to switch the genome from the chr20-21 version to the full thing, since the input VCF has variants from other contigs. The old test probably should not have worked (the reference genome was basically ignored then).

The biggest changes are in testSelect1.expected.txt and testStandardIndelEval.expected.txt. These changes are generally related to comp overlaps, and I am pretty sure all similar to my example above.

To state another way: the tool as-is builds a whitelist of sites, by first identifying any variant overlapping the driving variants (possibly limited by -L). If a multi-site variant overlaps one of those sites, it expands the ReferenceContext. This means if the original variant was a single-base variant at 1423282, but is it overlapped by a large indel starting at 1423281, then the reference context is expanded accordingly. The problem is that the tool as-is then iterates these expanded reference contexts, and asks for every variant in the Comp VCFs that overlaps (defined as starts within) these expanded intervals, not limiting itself to the Eval variants.

bbimber commented:
@cmnbroad OK, rebase + conflicts fixed. The changes in the Arguments looked pretty limited: just one File->GATKPath and another instance of File->Path.

To move forward: I left some debugging code that makes it relatively easy to explore the behavior change. See: https://github.com/broadinstitute/gatk/pull/6973/files#diff-510f106f1226c4aaef8bf898603f9c124a60065d56d08c8f333b42443fedc68cR235. I left the code that did the original re-query for variants. It logs when the old/new set of variants isnt equal, which you can also put a breakpoint on.

What's going to help you review this and/or discuss whether you think this is an acceptable difference?

cmnbroad commented:
@bbimber thanks for the rebase and update. I'll take a closer look at this today.

cmnbroad commented:
Ok, I think the main thing is to be able to explain/verify all of the tests where results changed:

- For the snpeff test, since the behavior on this branch seems more correct to me than master, I tried running the GATK4 test case inputs with GATK3, and it produces exactly the same results as this branch does. So I think that issue was introduced by the original GATK4 port, and is fixed in this branch.
- The rest of the tests with changed results don't seem to hit your breakpoint, though. So I think we need to figure out why they changed, and maybe also compare them with GATK3 (which can be a pain because the output format is slightly different).
-  As you mentioned, you changed the reference for testEvalTrackWithoutGenotypesWithSampleFields, which seems to have only affected the number of loci processed. So I'm unclear why that change was necessary. If the test truly should have been failing without this change, will it still fail if the change is reverted ? If not, can we fix it, and either way there should be a negative test for that case.

A few other general comments:

- I changed this PR to `draft` mode for now, which just better categorizes it for our internal workflow purposes. When its ready for a detailed code review we can remove the `draft` status.
- The `HashMap<FeatureInput<VariantContext>, HashMap<String, Collection<VariantContext>>>` can be wrapped in a class with just a couple of methods, so we don't have to manifest that long type all over the place.
- I know this PR still in an interim state, but passing the VariantWalker in as an argument to the comp methods  doesn't seem like a step forward to me. If we can't solve that problem completely in this PR (which is fine, I'm all for trying to contain this), are those changes necessary ? Perhaps that part should just wait for the next round.
- Any new classes/methods should use `final` for variables and parameters wherever applicable, and public classes and methods should have javadoc.
- Finally, I'm curious if you've tried any perf testing on this branch ? Is it better ?

bbimber commented:
OK, that's reasonable. I'll dig into the other test changes.

I can answer a few:

- Regarding passing the VariantWalker: I agree that's not an improvement by itself, but I would argue it's not that much different than it was. My plan is to pass a VariantEvalContext object, which would obscure any need to have knowledge of the walker. In an attempt to keep this PR simpler, I didnt complete that work. I do expect to make a second PR in relatively short order, once we get this resolved.

- With respect to testEvalTrackWithoutGenotypesWithSampleFields and the different reference: I think the issue is that the old version (master GATK branch) didnt validate as strictly. When switching to MultiVariantWalkerGroupedOnStart, the reference is required, and the tool will error if the contigs dont match. VariantEval on the master branch didnt really need the reference for anything, and was apparently more permissive if it didnt line up. It probably preferentially grabbed the dictionary from the VCF header.

I will look into those other questions

bbimber commented:
@cmnbroad OK, I tracked down two issues:

- The CpG and repeat status are dictates by the sequence spanning where ReferenceContext thinks it starts. This value is difference in the MultiVariantWalker version than before (basically the version on master can be expanded). I need to think about this. I half-fixed this, which restores the indel -related tests back to their original values (i.e. we no longer change behavior).
 
- I think in some cases the MultiVariantWalker will miss overlapping comp variants upstream of the eval.

bbimber commented:
@cmnbroad OK, considerable progress here. I was able to adjust behavior such that only two tests have changed behavior from GATK4/master. I think this is now correct. One instance of changed behavior is the Snpeff/overlap one we discussed above. The second is the one where we now provide the full genome as REF, not the truncated genome. I think this difference is justified since the tool now requires a reference, and the prior version was arguably too lenient on validation of contigs. Anyway, this branch now also removes by debugging code and comments. I think it is ready for a review.

To some other questions you had above:

1) The HashMap<FeatureInput<VariantContext>, HashMap<String, Collection<VariantContext>>> can be wrapped in a class with just a couple of methods, so we don't have to manifest that long type all over the place.

I realize that's non-optimal, but this isnt anything I introduced here. I would really like to keep this PR as limited as we can, and address some larger refactoring in a different PR, once we've migrated to MultiVariantWalkerGroupedOnStart.

2) I know this PR still in an interim state, but passing the VariantWalker in as an argument to the comp methods doesn't seem like a step forward to me. If we can't solve that problem completely in this PR (which is fine, I'm all for trying to contain this), are those changes necessary ? Perhaps that part should just wait for the next round.

As noted above, I'd like to propose this as iterative, with a second PR coming soon. I did this b/c it moved us toward not needing to pass around the walker. It minimizes the code that has access to the walker (as opposed to setting it after creating the instance of the Evaluator, etc. Yes, it exposes it for two methods, but those classes no longer hang on to it. I would like to ultimately remove this entirely.

3) To re-iterate testEvalTrackWithoutGenotypesWithSampleFields: the input file, noGenotypes.vcf, has a header dictionary with the full set of contigs, and a single variant from chr 1. Prior to this PR, the test executed and supplied b37_reference_20_21 as the reference FASTA. The variant in that VCF is from chr 1, not 20/21. The old code should have failed. It didnt, probably since it was preferentially taking the sequence dictionary from the VCF header and basically ignoring the FASTA. It didnt make much practical difference, but I believe my change here is right.

cmnbroad reviewed (commented):
@bbimber I didn't make it all the way through, but I'm checkpointing what I have so far. In addition to my other comments, there are still quite a few issues with access modifiers and missing finals, etc. in new code that can be cleaned up. I commented on some, but I'd suggest fixing those in all the new code.  Also, keep in mind that this PR is still blocked on a resolution (hopefully via https://github.com/broadinstitute/gatk/pull/4571) to the non-VariantEval test failures that are a side effect of putting the source name in each variant. Most of us will be out for the Broad end of year closure; I'll be back on 1/4, we'll have to resume then.

bbimber commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/varianteval/VariantEvalArgumentCollection.java:
```diff
@@ -0,0 +1,149 @@
+
+    // Public constants
+    final protected static String ALL_SAMPLE_NAME = "all";
+    final protected static String ALL_FAMILY_NAME = "all";
```
point taken, but a lot of this predates this PR

bbimber commented:
@cmnbroad just to check one thing: are you expecting that #4571 is actually going to get resolved outside of this PR? based on it not having any action in the last year I figured that PR was dead and I'd need to deal with anything related to that here? I had been focusing on VariantEval, but I will look into these other tests

bbimber commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantWalkerGroupedOnStart.java:
```diff
@@ -114,6 +116,25 @@ public final void apply(VariantContext variant, ReadsContext readsContext, Refer
+        Map<String, FeatureInput<VariantContext>> sourceMap = new HashMap<>();
+        getDrivingVariantsFeatureInputs().forEach(x -> sourceMap.put(x.getName(), x));
+
+        Map<FeatureInput<VariantContext>, List<VariantContext>> ret = new HashMap<>();
+        variants.forEach(vc -> {
+            FeatureInput<VariantContext> fi = sourceMap.get(vc.getSource());
+            if (fi == null) {
+                throw new GATKException.ShouldNeverReachHereException("Variant could not be mapped to a source: " + vc);
+            }
+
+            List<VariantContext> l = ret.getOrDefault(fi, new ArrayList<>());
+            l.add(vc);
+            ret.put(fi, l);
+        });
+
+        return ret;
```
ok

bbimber commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/varianteval/VariantEval.java:
```diff
@@ -466,66 +361,41 @@ protected EvaluationContext createEvaluationContext(final Set<Class<? extends Va
-                i = new SimpleInterval(i.getContig(), i.getStart(), vc.getEnd());
+    @Override
+    public void apply(List<VariantContext> variantContexts, ReferenceContext referenceContext, List<ReadsContext> readsContexts) {
+        Map<FeatureInput<VariantContext>, List<VariantContext>> variantMap = groupVariantsByFeatureInput(variantContexts);
```
OK. for the record, other subclasses of MultiVariantWalkerGroupedOnStart, like FilterAlignmentArtifacts and CombineGVCFs dont have final either. I assume everyone just let Intellij auto-generate the method, so it would inherit the argument pattern from MultiVariantWalkerGroupedOnStart...

cmnbroad commented:
@bbimber Yes, we discussed #4571 at a GATK team meeting a couple of weeks ago and we know what the issue is. Somebody (I think either @KevinCLydon or @jonn-smith) said they have a branch somewhere with a solution. So I wouldn't spend any time on those since hopefully we'll resurrect the branch with the fix and you can just rebase on that.

bbimber commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantWalkerGroupedOnStart.java:
```diff
@@ -114,6 +116,25 @@ public final void apply(VariantContext variant, ReadsContext readsContext, Refer
         lastCurrentVariantStart = variant.getStart();
     }
 
+    protected Map<FeatureInput<VariantContext>, List<VariantContext>> groupVariantsByFeatureInput(List<VariantContext> variants) {
```
I'll move to VariantEval

bbimber commented on src/main/java/org/broadinstitute/hellbender/engine/MultiVariantWalkerGroupedOnStart.java:
```diff
@@ -114,6 +116,25 @@ public final void apply(VariantContext variant, ReadsContext readsContext, Refer
 
+    protected Map<FeatureInput<VariantContext>, List<VariantContext>> groupVariantsByFeatureInput(List<VariantContext> variants) {
+        Map<String, FeatureInput<VariantContext>> sourceMap = new HashMap<>();
+        getDrivingVariantsFeatureInputs().forEach(x -> sourceMap.put(x.getName(), x));
```
ok

bbimber commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/varianteval/VariantEval.java:
```diff
@@ -126,125 +123,27 @@
-     */
-    @Argument(fullName="gold-standard", shortName = "gold", doc="Evaluations that count calls at sites of true variation (e.g., indel calls) will use this argument as their gold standard for comparison", optional=true)
-    public FeatureInput<VariantContext> goldStandard = null;
+    VariantEvalArgumentCollection variantEvalArgs = new VariantEvalArgumentCollection();
```
it is now private, but does have the getter. As noted, the PR isnt perfect, but neither was the original code.

bbimber commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/varianteval/VariantEval.java:
```diff
@@ -318,60 +213,60 @@ public void onTraversalStart() {
-            inputToNameMap.put(dbsnp.dbsnp, "dbsnp");
-            knowns.add(dbsnp.dbsnp);
+        variantEvalArgs.comps.addAll(variantEvalArgs.compsProvided);
+        variantEvalArgs.compsProvided.forEach(x -> inputToNameMap.put(x, x.hasUserSuppliedName() ? x.getName() : StandardArgumentDefinitions.COMPARISON_SHORT_NAME));
```
ok

bbimber commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/varianteval/VariantEvalArgumentCollection.java:
```diff
@@ -0,0 +1,149 @@
+
+    // Partitioning the data arguments
+    @Argument(shortName="select", doc="One or more stratifications to use when evaluating the data", optional=true)
+    protected ArrayList<String> SELECT_EXPS = new ArrayList<String>();
```
ok

bbimber commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/varianteval/VariantEvalArgumentCollection.java:
```diff
@@ -0,0 +1,149 @@
+     * Other sets can be specified with the -known-name (--known_names) argument.
+     */
+    @ArgumentCollection
+    protected DbsnpArgumentCollection dbsnp = new DbsnpArgumentCollection();
```
I did this and also a little more cleanup between VariantEval and VariantEvalArgumentCollection in expectation of introducing a new VariantEvalEngine after this PR is resolved

bbimber commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/varianteval/VariantEval.java:
```diff
@@ -126,125 +123,27 @@
-     * The variant file(s) to evaluate.
-     */
-    @Argument(fullName="eval", shortName = "eval", doc="Input evaluation file(s)", optional=false)
-    public List<FeatureInput<VariantContext>> evals;
```
Yes - this was trying to make a compromise and move toward the goal of creating the standalone VariantEvalEngine.

What about letting the ArgumentCollection encapsulate some of logic around creating DrivingVariants? In theory one might want different walkers based around this, and it would be helpful to have one piece of code that does that? Or do you think the better pattern is to make the VariantEvalEngine class that takes the ArgumentCollection in the constructor, and then perhaps that provides methods to set up DrivingVariants?

bbimber commented:
@cmnbroad OK - I think I addressed your review and also added a sizable amount of 'final', including to pre-existing code. 

I did not do the refactor of eval/comp/gold-standard from VariantEvalArgumentCollection back into VariantEval. I see your point and am not totally opposed to it, but put a question on that thread.

bbimber commented:
@cmnbroad After thinking about this I went ahead and created VariantEvalEngine. Doing this in one PR will simplify some of the sticking points around what is a final change vs. what it expected to be fixed later. With this change, the goal is to strip most logic from VariantEval into the engine. This engine can be constructed with a VariantEvalArgumentCollection, and any kind of GATKTool as the owner. I tried to minimize the amount of context the VariantEvalEngine needed to hang on to. This means all the child classes have visibility on the VariantEvalEngine, but are no longer directly exposed to either the walker class or the argument collection. 

All the logic around gathering the arguments to form DrivingVariants is moved to a static method in VariantEvalEngine. 

I also rebased and fixed conflicts.

bbimber commented:
Final bit of information: see the proposed change I have in GATKTool related to default ReadFilters. Previously VariantEval overrode getPluginDescriptors() to prevent the default read filters from getting applied. The reason for this is b/c these included SampleReadFilter, which has an argument named 'sample'. As a result, that conflicted with our tool. Because VariantEval doesnt actually use reads, there's no reason to be including getDefaultReadFilters(), unless I'm missing something.

bbimber commented:
This will definitely not be part of this PR, but I wonder if you'd be willing to comment on this: it would be quite helpful if I could get VariantEval to run in a scatter/gather mode, in which I submit jobs to the cluster, each saves results for an internal and then I combine them. One way to make this work would be for VariantEvalEngine to save the StratificationManager (or some version of this) to disk, perhaps serialized to JSON with Jackson. There would need to be a separate tool that takes N serialized StratificationManager objects, loads then, merges, and then writes the actual report. It's not completely implemented, but there is already a concept of combineStrats() in StratificationManager. It doesnt work complete, but this doesnt seem all that far from being able to take two StratificationManager objects and returning the merged form.  Is this a reasonable idea?

bbimber commented:
@cmnbroad Hope you had a good break. As you might have seen above, I refactored out VariantEvalEngine, which I think will address some of the problems, like passing the walker around, which you didnt like in the PR. 

I might not have as good an eye over tests as you, but I believe the failures mostly relate to the change to set 'source' on the VCs. As you probably know, that means it's passed into the resulting VCF, and will change test expectations in some cases. You said someone at GATK was looking into that, but I would be happy to take a stab if your team has ideas on how to address this.  If I'm missing test failures beyond that I can take a look.

cmnbroad reviewed (commented):
@bbimber This is looking much better - thanks for factoring out the eval engine. That helped a lot. There is still a dependency on the tool, but I think its easily remedied - see my inline comments/questions before I do a more thorough review.

I didn't look at all of the test failures, but every one I did look at was due to failed VariantContext equality arising from the change to set the source, so I wouldn't spend any time on those. We have a resolution for it but it involves getting a couple of other PRS in. We are working on getting those resurrected and merged.

bbimber commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/varianteval/VariantEvalEngine.java:
```diff
@@ -0,0 +1,940 @@
+        }
+    }
+
+    public static List<FeatureInput<VariantContext>> getFeatureInputsForDrivingVariants(VariantEvalArgumentCollection variantEvalArgs) {
```
yes, good point

bbimber commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/varianteval/VariantEvalEngine.java:
```diff
@@ -0,0 +1,940 @@
+
+    public VariantEvalEngine(VariantEvalArgumentCollection variantEvalArgs, GATKTool owner, SAMSequenceDictionary samSequenceDictionaryForDrivingVariants, @Nullable Collection<String> samples, Logger logger) {
+        this.variantEvalArgs = variantEvalArgs;
+        this.owner = owner;
```
ok

bbimber replied:
Note: I dont know how intentional you phrased that, but I'm proposing to pass the value of getTraversalIntervals(), so we dont duplicate that logic

bbimber commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/varianteval/VariantEvalEngine.java:
```diff
@@ -0,0 +1,940 @@
+    private ReferenceSequenceFile ancestralAlignments = null;
+
+    // The set of all possible evaluation contexts
+    StratificationManager<VariantStratifier, EvaluationContext> stratManager;
```
How about a protected getter for now?

bbimber commented:
@cmnbroad OK, changes added

cmnbroad commented:
This PR is dependent on https://github.com/broadinstitute/gatk/pull/7021, which in turn will require  https://github.com/broadinstitute/gatk/pull/4571, plus one additional PR to change all of the tests to use the new  variant compare methods.

bbimber commented:
@cmnbroad OK - is there a person assigned to those and a timeline? can I do anything that will accelerate them?

bbimber commented:
Also, and again not part of this PR, does the concept of trying to run VariantEvalEngine over an interval, serializing its state to disk somehow (which would involve doing something with StratificationManager), and then writing a tool that loads/aggregates those per-interval objects seem reasonable as a way to support scatter/gather?

cmnbroad commented:
I mentioned to the team yesterday that you were willing to help with moving those along. As for scatter/gather, based on the code structure I assume that GATK3 used the stratification manager/combine methods to do something similar to what you describe, but I'm not very familiar with the gatk3 code base, so its only an assumption.

bbimber commented:
OK. StratManager is about 80% wired to do the merging this would require. It would require me to implement some existing methods, like VariantEvaluator.combine(), but the basic idea exists today in StratificationManager.combineStrats(). I'd probably propose to expose this via VariantEvalEngine by adding saveToDisk(File targetFile), and restore(List<File> serializedStratManagers) methods. I'd propose to keep interaction with this process protected and only expose the save/restore methods. Alternately, I could expose save, restore (single-object) and combine methods.

Regarding actually saving the state of StratificationManager to disk: some form of Jackson-based serialization would probably be easiest. My initial thought would be to make a new class specifically for serializing: SerializedStratificationManager. This would cache the relevant information and be the object that is serialized/deserialized. I would add a new constructor to StratificationManager that accepts this object and restores the state within StratificationManager. I think this could be all be kept internal to GATK.

Does that broad outline seem like a concept you might accept in GATK?

cmnbroad commented:
@bbimber I'd suggest creating a separate issue ticket with that proposal, so I and others can chime in on it there.

cmnbroad reviewed (changes requested):
Checkpointing my still incomplete partial review.

Also, @bbimber can you confirm that the changes in this PR sufficiently enables you to switch VariantQC from subclassing the tool to using the new `VariantEvalEngine` and `VariantEvalArgumentCollection` classes directly ? (You probably know how to use a local GATK build to try that out, but if not I can tell you how to do that - its fairly easy). It would be a really useful validation of these changes.

bbimber commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/varianteval/evaluators/IndelLengthHistogram.java:
```diff
@@ -44,7 +43,7 @@ private void initializeCounts(int size) {
 
     @Override
-    public void finalizeEvaluation() {
+    public void finalizeEvaluation(final VariantEvalEngine engine) {
```
ok

bbimber commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/varianteval/evaluators/VariantEvaluator.java:
```diff
@@ -19,26 +16,22 @@ protected VariantEvaluator() {
-    public void initialize(VariantEval walker) {
-        this.walker = walker;
-    }
+    public void initialize(final VariantEvalEngine engine) {
```
Yes, I see your point on pushing that into the constructor. I see GATK has a ClassUtils set of helpers, but I dont see anything that helps you find a constructor by arguments. I assume you want me to do something like this in EvauationContext?

Current:

        for ( final Class<? extends VariantEvaluator> c : evaluationClasses ) {
            final VariantEvaluator eval = ClassUtils.makeInstanceOf(c);
            //c.getDeclaredConstructor(VariantEvalEngine.class)
            if ( doInitialize ) eval.initialize(engine);
            evaluationInstances.add(eval);
        }

New pattern:

        for ( final Class<? extends VariantEvaluator> c : evaluationClasses ) {
             //dont do this:
            //final VariantEvaluator eval = ClassUtils.makeInstanceOf(c);
            //instead get constructor along these lines (this is obviously not complete). 
             c.getDeclaredConstructor(VariantEvalEngine.class)
            
            if ( doInitialize ) eval.initialize(engine);
            evaluationInstances.add(eval);
        }


Only reason I ask is that I dont see similar examples on reflection like that in GATK4 and wasnt sure if there was a different pattern to use.

bbimber commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/varianteval/VariantEvalEngine.java:
```diff
@@ -0,0 +1,937 @@
+        return variantEvalArgs.ploidy;
+    }
+
+    public Logger getLogger() {
```
ok

bbimber commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/varianteval/VariantEvalEngine.java:
```diff
@@ -0,0 +1,937 @@
+
+    // No args constructor for unit testing only
+    @VisibleForTesting
+    protected VariantEvalEngine(Logger logger) {
```
ok

bbimber commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/varianteval/VariantEvalEngine.java:
```diff
@@ -0,0 +1,937 @@
+        return Collections.unmodifiableSet(standardEvaluatorNames);
+    }
+
+    public FeatureInput<Feature> getKnownCNVsFile() {
```
ok

bbimber commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/varianteval/evaluators/VariantEvaluator.java:
```diff
@@ -1,14 +1,11 @@
-    private VariantEval walker;
     private final String simpleName;
 
     protected VariantEvaluator(String simpleName) {
```
ok

bbimber replied:
When I checked out VariantQC, I see I created it to support that tool.  I added a comment to that effect - any concerns about keeping it? I could think about other ways to approach this. The easiest would be to add a setSimpleName(), but I'm not sure that's really better than this.

bbimber commented:
@cmnbroad Depending on whether you like my changes in VariantEvalEngine.createVariantStratifier() and createVariantEvaluator() or not, I think I just committed changes for all your comments.

I actually have not built another project using a local GATK JAR - would you mind saying a little more on that? I agree that would be important here to make sure our VariantQC tool can work with this VariantEvalEngine.

bbimber commented:
Actually, as far as building with my local GATK, I remembered you can point to local JARs. Is the easiest solution to update my other project to point to the local GATK jar, like so:

dependencies {
    implementation files('libs/something_local.jar')
}

cmnbroad commented:
@bbimber I usually do it like this:

- run `./gradlew clean install printVersion` on gatk (this will build and install gatk into the local maven repo, and display a SNAPSHOT version number of the installed artifact)
- add `mavenLocal()` to the repository list (in your case, to the VariantQC gradle file)
- update the gatk version in your gradle file to the use SNAPSHOT version printed out by the command above
- do a clean build

bbimber commented:
@cmnbroad OK. I got there with a slightly hackier way (put the JAR in ./lib of my project and added it in gradle), but I will check out your method. Do I need some kind of publishToMavenLocal() step from the GATK project, or is it somehow automatically visible?

cmnbroad commented:
The `./gradlew clean install printVersion` command I mentioned publishes it to the local maven repo. Then you need to reference that version from your own build.gradle. Also beware when doing this iteratively that the local snapshot version # is based on the current head commit hash. In my experience, once gatk is published to local maven under a given version, if you want to make a change, you'll need to make a commit to force the version number to update, and then rerun the command above, and update your VariantQC to the new version. Otherwise the local repo won't be updated since it will appear up to date. Hope that makes sense.

bbimber commented:
ok, got it. sorry, i missed 'install' in that command. my initial impression is that VariantQC will be able to adapt fine to VariantEvalEngine. I wrote VariantEvalEngine with this is mind, but it's good to formally test it.

bbimber commented:
@cmnbroad With two changes I put in this morning, I have formally confirmed this will work with VariantQC. 

Any updates on those two blocking PRs?

bbimber commented:
@cmnbroad I thought I'd check in. Any further thoughts on this, and/or updates on those two blocking PRs?  Do you expect to be doing additional review of this code, or do you feel OK with the changes as-is (obviously conditional on those two other PRs)?

cmnbroad commented:
@bbimber I think its very close, but I still need to do a final pass (I'm out the rest of the week so it will be early next week). We haven't forgotten about the other PRs, we're just slow. The restored constructor seems fine.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/varianteval/VariantEvalEngine.java:
```diff
@@ -0,0 +1,922 @@
+        drivingVariantSourceMap = new HashMap<>();
+        variantEvalArgs.getFeatureInputsForDrivingVariants().forEach(x -> drivingVariantSourceMap.put(x.getName(), x));
+
+        if (!skipInitialize) {
```
I thought we were getting close, but this deferred initialization state seems complicating - whats the reason for needing it (both here, and in the evaluators, which still have an "initialize" method) ? Is there any alternative ?

Depending on the answer, we'll need to rationalize these constructors - having one that controls initialization and one that doesn't seems problematic. There are 3 constructors now, each of which leaves the object in a slightly different non-obvious state. We really need to get rid of the boolean if at all possible.

bbimber replied:
The reason for this is listed in the comment above that method: it was to make this work with our VariantQC tool. I need to delay creation of EvaluationContext for some things our tool does. However, I took a look at my code and I think I have a solution that doesnt need this, so I removed it here.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/varianteval/evaluators/VariantEvaluator.java:
```diff
@@ -1,41 +1,42 @@
-    public void initialize(VariantEval walker) {
-        this.walker = walker;
+    public void initialize() {
+
```
I thought passing the engine to constructors would eliminate the need for this initialize method. Can it be removed ?

bbimber replied:
VariantSummary is an example of one class using this to initialize variables. It doesnt seem crazy to me to keep this? In theory this class could store some internal "hasInitialized" boolean, and lazily do this work in the update() method, but initialize() doesnt seem that harmful to exist? Example below:

```
    @Override
    public void initialize() {
        this.knownCNVs = getEngine().getVariantEvalArgs().getKnownCNVsFile();

        nSamples = getEngine().getSampleNamesForEvaluation().size();
        countsPerSample = new TypeSampleMap(getEngine().getSampleNamesForEvaluation());
        transitionsPerSample = new TypeSampleMap(getEngine().getSampleNamesForEvaluation());
        transversionsPerSample = new TypeSampleMap(getEngine().getSampleNamesForEvaluation());
        allVariantCounts = new TypeSampleMap(getEngine().getSampleNamesForEvaluation());
        knownVariantCounts = new TypeSampleMap(getEngine().getSampleNamesForEvaluation());
        depthPerSample = new TypeSampleMap(getEngine().getSampleNamesForEvaluation());
    }

```

bbimber replied:
I thought about this and I agree - initialize() doesnt really need to exist. I moved the two classes that were actually using it to do that work in their constructors.

cmnbroad reviewed (commented):
@bbimber Sorry again for the delay in this - a couple of questions about the delayed initialization.

bbimber commented:
@cmnbroad please see comments to review above. note, i just squashed/rebased as well.

any word on those two related PRs?

cmnbroad commented:
@bbimber This this looks pretty good now -  I don't have any new substantive comments. We just need to get the other PRs in and the tests passing and this should be good to go.

There is one outstanding thing that would be nice, but I'll leave it up to you as to whether you want to address it. `VariantStratifier` and it's subclasses still have `initialize` methods that take no arguments. As far as I can tell, this is unnecessary now that they are required to have a constructor that take a `VariantEvalEngine`. If you feel so inclined, the `initialize` content can be moved into the constructors, and the initialize method, and call to it, can be eliminated (BTW, the constructors in those classes should be moved to be after the variable declarations).

Overall, though, I think this PR makes a pretty big improvement to the state of things. Thanks for doing it.

bbimber commented:
@cmnbroad Is there anything that can be done to accelerate those other PRs? As i said before I'd be willing to take a stab at updating anything. We're very keen to get this merged so we can update our VariantQC tool...

VariantStratifier.initialize() is removed

cmnbroad commented:
@bbimber The rate limiting step on the other PRs, at the moment anyway, is code review, so there isn't much else we can do  until thats completed. Also, I don't see the commit where VariantStratifier.initialize() was removed (maybe you haven't pushed it yet ?)

bbimber commented:
@cmnbroad yeah, didnt push the changes - trying to do too many things at once. they are added, with a squash/rebase for good measure.

bbimber commented:
@cmnbroad Hey, just wanted to check in - i dont suppose there are any updates on those blocking PRs?

bbimber commented:
@cmnbroad Hey, just wanted to check in again. I posted directly on the other blocking PRs, but I want to renew my offer to help on that code if it would help get those complete, so we can finally close this out.

bbimber commented:
@cmnbroad On these blocking PRs. The core point is to provide a way to link FeatureInput to VariantContext. The existing PRs stash FeatureInput name as the VariantContext source. I dont know if there are internal deliberations or anything holding up getting those done, or blocking my offer to try to implement that.

However, I wonder what you think about a possible alternate approach. The driving use for this feature is MultiVariantWalkers and their subclasses. Another way to tackle this might be to hook into MultiVariantDataSource. One could create a wrapper class:

```
    public class VariantContextWithSource {
        VariantContext variant;
        FeatureInput source;
    }

```
And the iterator layer would wrap the VariantContext objects with this class, making a non-text link to FeatureInput. And perhaps either a) apply() would get called with this wrapper instead of VariantContext, or b) MultiVariantWalker might be able to get converted into MultiVariantWalker<T> where T is the class provided to apply(). I havent fully explored this, but it might be a more convenient way to share implementations.

I could try to make this more concrete and propose something, but I dont want to spend work if the concept of wrapping VariantContext into something like VariantContextWithSource isnt going to go anywhere.

cmnbroad commented:
@bbimber Populating the VariantContext source field is something we want anyway, so introducing additional classes/complexity as an alternative doesn't seem like a good path.

Also, as I've mentioned before, the blocking issue on the other PRs is internal code review.

bbimber commented:
@cmnbroad OK, thanks. I appreciate your help and time on this. It's sometimes difficult to evaluate what's going on with the review with relatively little information available out here.

bbimber commented:
@cmnbroad I thought I'd check in on these PRs. As before, if there is anything I can do that would help accelerate getting those blocking PRs finalized I am more than happy to take on parts of that work.

cmnbroad commented:
@bbimber I do apologize again for the long timeframe on this. After more team discussion today, and given the amount of work we think it will take to get https://github.com/broadinstitute/gatk/pull/4571 merged, we probably need to consider an alternative path forward for this PR. One possibility is to create another version of `MultiVariantWalkerGroupedOnStart` that effectively does what https://github.com/broadinstitute/gatk/pull/4571 does (might require a special version of `MultiVariantDataSource` as well), and use that just for `VariantEval`.  Although not ideal, it might be the best compromise to preserve all the work you've done in this PR and its get it merged in the near future.

bbimber commented:
@cmnbroad Thanks for the reply. Could you help me understand a little better the complexity of #4571? What about adding this addition is going to take a lot of changes to address?

As far as a second workaround solution, hooking into MultiVariantDataSource was exactly what I was proposing above (see idea WRT "VariantContextWithSource"). I agree that's has aspects that could be a little clunkier, but I think it's possible to either have a base class for MultiVariantWalkerGroupedOnStart, or use generics in a way that allows most of the MultiVariantWalkerGroupedOnStart code to be re-used. I have not tried to code anything up yet.

bbimber commented:
Another permutation on this: i dont know if you'll like hooking into the codec, but one could wrap the codec and create a FeatureInputAwareVariantContext. This would go into FeatureDataSource, and would potentially allow Features to be 'aware' of their owning FeatureInput. The code below is not final and needs work, but this or the patch might give an idea:

[FeatureInputAwareVariantContext.patch.txt](https://github.com/broadinstitute/gatk/files/6346205/FeatureInputAwareVariantContext.patch.txt)


```
    public static class CodecWrapper<FEATURE_TYPE extends Feature, SOURCE> implements FeatureCodec<FEATURE_TYPE, SOURCE>
    {
        private final FeatureCodec<FEATURE_TYPE, SOURCE> childCodec;
        private final FeatureInput<FEATURE_TYPE> featureInput;

        public CodecWrapper(FeatureCodec<FEATURE_TYPE, SOURCE> childCodec, FeatureInput<FEATURE_TYPE> featureInput)
        {
            this.childCodec = childCodec;
            this.featureInput = featureInput;
        }

        @Override
        public Feature decodeLoc(SOURCE source) throws IOException {
            return childCodec.decodeLoc(source);
        }

        @Override
        public FEATURE_TYPE decode(SOURCE source) throws IOException {
            FEATURE_TYPE feature = childCodec.decode(source);

            //Either look for marker class or otherwise poke in FeatureInput here:
            if (feature instanceof VariantContext)
            {
                feature = new FeatureInputAwareVariantContext(feature, featureInput);
            }

            return feature;
        }

        @Override
        public FeatureCodecHeader readHeader(SOURCE source) throws IOException {
            return childCodec.readHeader(source);
        }

        @Override
        public Class<FEATURE_TYPE> getFeatureType() {
            return childCodec.getFeatureType();
        }

        @Override
        public SOURCE makeSourceFromStream(InputStream bufferedInputStream) {
            return childCodec.makeSourceFromStream(bufferedInputStream);
        }

        @Override
        public LocationAware makeIndexableSourceFromStream(InputStream inputStream) {
            return childCodec.makeIndexableSourceFromStream(inputStream);
        }

        @Override
        public boolean isDone(SOURCE source) {
            return childCodec.isDone(source);
        }

        @Override
        public void close(SOURCE source) {
            childCodec.close(source);
        }

        @Override
        public boolean canDecode(String path) {
            return childCodec.canDecode(path);
        }
    }

    public static interface FeatureInputAware<FEATURE_TYPE extends Feature>
    {
        public FeatureInput<FEATURE_TYPE> getFeatureInput();
    }

    public static class FeatureInputAwareVariantContext extends VariantContext implements FeatureInputAware<VariantContext>
    {
        private FeatureInput<VariantContext> featureInput;

        public FeatureInputAwareVariantContext(VariantContext parent, FeatureInput<VariantContext> featureInput)
        {
            super(parent);
            this.featureInput = featureInput;
        }

        @Override
        public FeatureInput<VariantContext> getFeatureInput() {
            return featureInput;
        }
    }


```

cmnbroad commented:
Ideally we'd minimize changes to core engine classes, and instead create new specialized subclasses (like a subclass of MultiVariantWalkerGroupedOnStart) that implement the desired behavior. The main thing is that we need something that only affects `VariantEval` - anything that affects other tools would run into the same issues as #4571 - specifically that tests that use VariantContext equality to validate results will start to fail because of the presence of source name in the actual values.

bbimber commented:
I appreciate the desire for minimal changes, but I would point out that tying the VariantContext to the source FeatureInput is likely to be a somewhat common need for MultiVariantWalkers. I realize you have a lot of existing example that dont need this capability. While I started this using VariantEval/VariantQC, I more recently tried to port CombineVariants to GATK4 and hit a similar roadblock. I have one or two other lab-specific walkers that would benefit from using the iteration pattern of MultiVariantWalkerGroupedOnStart, but also need some ability to retain the FeatureInput->VariantContext link. I believe GATK3 retained this. It would be nice to at least make this a capability available to all MultiVariantWalkerGroupedOnStart walkers.

My suggestion above about making a FeatureInputAwareVariantContext wasnt necessarily meant to be introduced into every class. Would something conceptually like this pass GATK if I could introduce it more surgically, perhaps injected into MultiVariantDataSource alone?

bbimber commented:
Again, it's difficult to hold these discussions over issues and understand exactly where the blockers are, but one argument in favor of the wrapper class (FeatureInputAwareVariantContext), instead of re-purposing the source field, is that we can a) retain a reference to the actual FeatureInput class, and b) this should not change existing tests.

cmnbroad commented:
If I understand what you're proposing, the tool code would have to use `instanceof` checks and typecasts to take advantage of this feature (unless `FeatureInputAwareVariantContext` was exposed in the `apply` method). Other reviewers may feel differently, but using the source field (which I don't view as "repurposing" since I think its aligned with the intent of that field) seems cleaner. Up to you if you want to submit a PR with your proposed change to see if its viable.

bbimber commented:
@cmnbroad If this approach was used, I think it would make sense to make the MultiVariantWalker use generics in some form, so we call apply() on either VariantContext or FeatureAwareVariantContext. 

Nonetheless, I think I have two other simpler proposals, either: #7219  or #7220. Each of these are limited in the scope of change, and should only populate the source of the VC in limited situations.

bbimber commented:
@cmnbroad This is rebased onto MultiVariantDataSource2. I did not squash since I dont know whether you want all these changes as one PR, or to check in MultiVariantDataSource2 stand-alone, and then rebase/merge this one.

bbimber commented:
@cmnbroad rebasing is done. To summarize changes since your last review:

- I backed out the earlier changes to FeatureInput/FeatureDataSource in favor of those from #7219 
- I dont entirely know why this didnt hit before, but I made an update to VariantStratifiers to make tests pass. See: https://github.com/broadinstitute/gatk/pull/6973/commits/1569a909d3dc2301337e46441cc0cd969843c8d1. The gist is that we now instantiate those classes and pass VariantEvalEngine. Two of these classes had validation in their constructors, and could throw CommandLineException. This was getting caught and re-thrown as GATKException with the misleading message "Problem making an instance of ....". This proposal is to make a separate VariantStratifier.validateArgs() method, with a default no-op validation, and to call this only after instantiation. 

VariantEval tests pass locally for me. With luck, tests will pass here.

cmnbroad commented:
Hm - I don't think we can take that last change. Theres not much use in validating args after they've been used by the constructors. Let me see if there is an alternative.

bbimber commented:
> Hm - I don't think we can take that last change. Theres not much use in validating args after they've been used by the constructors. Let me see if there is an alternative.

Well in the prior commit I just caught CommandLineException (https://github.com/broadinstitute/gatk/pull/6973/commits/814839f498cda8ce627a47229d77fb6cac7ca6e0), but this seemed hacky. 

Why cant there be a separate method to create the class and validate args? 

There is VariantEvalEngine.validateAndInitialize(), where some classes do analogous checks.

cmnbroad commented:
The constructors (some of them anyway) use the args during constructor execution, so validating them after the constructor is done executing is too late. Anyway, if you revert that commit so the validation happens as before, and instead add the following to the catch block in `VariantEvalEngine::createClass`, they will get propagated as expected:

```
            if (e.getCause() instanceof CommandLineException) {
                // failures due to argument validation should be propagated
                throw new CommandLineException(e.getCause().getMessage(), e.getCause());
            }
            throw new GATKException("Problem making an instance of " + clazz + " Do check that the class has a constructor that accepts VariantEvalEngine", e);
```

bbimber commented:
> The constructors (some of them anyway) use the args during constructor execution

That's a theoretical problem but in this case they dont. Anyway, I restored my earlier code to special-case CommandLineException and throw that. 

You may have seen this, but there was a failure in AlleleFrequencyQCTest, which I addressed here: https://github.com/broadinstitute/gatk/pull/6973/commits/8a9f5db6eccbc82600f8ec46e7656bcf92bedf5b. The general change is to stop stashing two variables locally within AlleleFrequency, which allows AlleleFrequencyQC to set their values later.

cmnbroad commented:
Yeah, apparently some of these tests weren't running before (maybe because of the VariantContext equality failures caused by the global source setting) ? Anyway, there is an old [ticket](https://github.com/broadinstitute/gatk/issues/6997) to update AlleleFrequencyQC to embrace `VariantEvalEngine` once this PR goes in.

cmnbroad commented:
@bbimber VariantEvalIntegrationTest line 502 needs to change from `CommandLineException.BadArgumentValue.class` to `CommandLineException.class` to reflect the change in the catch block we made earlier.

bbimber commented:
@cmnbroad yes, was literally just checking that in. I agree it's probably better to rethrow and preserve the stack than preserving the original class like my first commit did.

Do you expect that so few of the travisci suites seem to actually run these integration tests? only three actually failed?

cmnbroad commented:
@bbimber Yes, most of those travis jobs are unique and do not overlap in terms of coverage. The unit and integration tests are each run in two separate jobs, one on Java 8 and one on Java 11, but the other jobs run python tools, cloud tests, WDL tests, docker image tests, etc. and do not overlap.

bbimber commented:
OK, I addressed one more AlleleFrequencyQC issue (https://github.com/broadinstitute/gatk/pull/6973/commits/6961f757cd3c20470f3b42b473aa948593e537b1).

bbimber commented:
@cmnbroad finally got a clean test run. what do you think about the PRs? should i squash/rebase #7219 (or maybe you'd be OK doing that via github on approval)? with that done, i will then squash/rebase this PR?

cmnbroad commented:
You can rebase on master now (I merged the other branch) - I'll squash on merge.

bbimber commented:
@cmnbroad OK, done

bbimber commented:
@cmnbroad it seems like we hit about 5 jobs with docker 'toomanyrequests' problems. how do you want to handle that?

cmnbroad commented:
@bbimber I'll be AFK much of today and tomorrow, but I'll restart these tonight when hopefully there will be less competition for the pull limit (master is failing at the moment as well). It may take a couple of days to get this through but I will keep an eye on it.

bbimber commented:
@cmnbroad it appears we have clean tests now?

cmnbroad reviewed (approved):
Finally this one looks ready. Thanks @bbimber.

bbimber commented:
@cmnbroad Thank you for all your time on this!
